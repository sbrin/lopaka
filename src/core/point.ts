import {Rect} from './rect';

/**
 * Point2d class
 * author: @deadlink
 */
function isArray(obj: any): boolean {
    return Object.prototype.toString.call(obj) === '[object Array]';
}
const x: symbol = Symbol('x');
const y: symbol = Symbol('y');
export class Point {
    constructor();
    constructor(point: Point);
    constructor(v: number[]);
    constructor(v: number);
    constructor(x: number, y: number);
    constructor(...args: any[]) {
        if (args.length) {
            if (args.length == 1) {
                if (args[0] instanceof Point) {
                    this[x] = args[0].x;
                    this[y] = args[0].y;
                } else if (isArray(args[0])) {
                    this[x] = args[0][0];
                    this[y] = args[0][1];
                } else {
                    this[x] = args[0];
                    this[y] = args[0];
                }
            } else {
                this[x] = args[0];
                this[y] = args[1];
            }
        } else {
            this.xy = [0, 0];
        }
    }
    get x(): number {
        return this[x];
    }
    set x(v: number) {
        this[x] = v;
    }
    get y(): number {
        return this[y];
    }
    set y(v: number) {
        this[y] = v;
    }
    get xy(): number[] {
        return [this[x], this[y]];
    }
    set xy(v: number[]) {
        this[x] = v[0];
        this[y] = v[1];
    }
    add(v: number): Point;
    add(v: number[]): Point;
    add(x: number, y: number): Point;
    add(v: Point): Point;
    add(...args: any[]): Point {
        if (args.length == 1) {
            if (args[0] instanceof Point) {
                this[x] += args[0].x;
                this[y] += args[0].y;
            } else if (isArray(args[0])) {
                this[x] += args[0][0];
                this[y] += args[0][1];
            } else {
                this[x] += args[0];
                this[y] += args[0];
            }
        } else {
            this[x] += args[0];
            this[y] += args[1];
            return this;
        }
        return this;
    }
    subtract(v: number): Point;
    subtract(v: number[]): Point;
    subtract(x: number, y: number): Point;
    subtract(v: Point): Point;
    subtract(...args: any[]): Point {
        if (args.length == 1) {
            if (args[0] instanceof Point) {
                this[x] -= args[0].x;
                this[y] -= args[0].y;
            } else if (isArray(args[0])) {
                this[x] -= args[0][0];
                this[y] -= args[0][1];
            } else {
                this[x] -= args[0];
                this[y] -= args[0];
            }
        } else {
            this[x] -= args[0];
            this[y] -= args[1];
            return this;
        }
        return this;
    }
    multiply(v: number): Point;
    multiply(v: number[]): Point;
    multiply(x: number, y: number): Point;
    multiply(v: Point): Point;
    multiply(...args: any[]): Point {
        if (args.length == 1) {
            if (args[0] instanceof Point) {
                this[x] *= args[0].x;
                this[y] *= args[0].y;
            } else if (isArray(args[0])) {
                this[x] *= args[0][0];
                this[y] *= args[0][1];
            } else {
                this[x] *= args[0];
                this[y] *= args[0];
            }
        } else {
            this[x] *= args[0];
            this[y] *= args[1];
            return this;
        }
        return this;
    }
    divide(v: number): Point;
    divide(v: number[]): Point;
    divide(x: number, y: number): Point;
    divide(v: Point): Point;
    divide(...args: any[]): Point {
        if (args.length == 1) {
            if (args[0] instanceof Point) {
                this[x] /= args[0].x;
                this[y] /= args[0].y;
            } else if (isArray(args[0])) {
                this[x] /= args[0][0];
                this[y] /= args[0][1];
            } else {
                this[x] /= args[0];
                this[y] /= args[0];
            }
        } else {
            this[x] /= args[0];
            this[y] /= args[1];
            return this;
        }
        return this;
    }
    min(point: Point): Point {
        return new Point(Math.min(this[x], point.x), Math.min(this[y], point.y));
    }
    max(point: Point): Point {
        return new Point(Math.max(this[x], point.x), Math.max(this[y], point.y));
    }
    abs(): Point {
        this[x] = Math.abs(this[x]);
        this[y] = Math.abs(this[y]);
        return this;
    }
    clone(): Point {
        return new Point(this);
    }
    round(): Point {
        this[x] = Math.round(this[x]);
        this[y] = Math.round(this[y]);
        return this;
    }
    floor(): Point {
        this[x] = Math.floor(this[x]);
        this[y] = Math.floor(this[y]);
        return this;
    }
    ceil(): Point {
        this[x] = Math.ceil(this[x]);
        this[y] = Math.ceil(this[y]);
        return this;
    }
    swap(): Point {
        const t = this[x];
        this[x] = this[y];
        this[y] = t;
        return this;
    }
    sign(): [number, number] {
        return [Math.sign(this[x]), Math.sign(this[y])];
    }
    distanceTo(p: Point): number {
        return Math.hypot(this.x - p.x, this.y - p.y);
    }
    normalize(): Point {
        return this.clone().divide(Math.sqrt(this[x] * this[x] + this[y] * this[y]));
    }
    boundTo(rect: Rect): Point {
        this[x] = Math.min(Math.max(this[x], rect.x), rect.w);
        this[y] = Math.min(Math.max(this[y], rect.y), rect.h);
        return this;
    }
    equals(p: Point): boolean {
        return this[x] == p.x && this[y] == p.y;
    }
    // todo
    pack(): number[] {
        return this.xy;
    }
    static unpack(data: number[]): Point {
        return new Point(data);
    }
}
