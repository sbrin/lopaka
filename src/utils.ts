function bitswap(b) {
    b = ((b & 0xf0) >> 4) | ((b & 0x0f) << 4);
    b = ((b & 0xcc) >> 2) | ((b & 0x33) << 2);
    b = ((b & 0xaa) >> 1) | ((b & 0x55) << 1);
    return b;
}

export function readFileAsync(file) {
    return new Promise((resolve, reject) => {
        let reader = new FileReader();

        reader.onload = () => {
            resolve(reader.result);
        };
        reader.onerror = reject;
        reader.readAsDataURL(file);
    });
}

export async function loadImageAsync(src): Promise<HTMLImageElement> {
    return new Promise((resolve, reject) => {
        let img = new Image();
        img.crossOrigin = 'anonymous';
        img.src = src;
        img.onerror = reject;
        img.onload = () => resolve(img);
    });
}

export function hexToRgb(hexColor: string) {
    const hex = hexColor.replace('#', '');
    const r = parseInt(hex.substring(0, 2), 16);
    const g = parseInt(hex.substring(2, 4), 16);
    const b = parseInt(hex.substring(4, 6), 16);
    return {r, g, b};
}

export function packedHexColor(color: string) {
    const c = packColor565(color);
    return `0x${c.toString(16).toUpperCase()}`;
}

export function packColor565(hexColor: string) {
    const rgb = hexToRgb(hexColor);
    return ((rgb.r >> 3) << 11) | ((rgb.g >> 2) << 5) | (rgb.b >> 3);
}

export function imgDataToXBMP(
    imgData: ImageData,
    xStart: number,
    yStart: number,
    width: number,
    height: number,
    swapBits: boolean = false
): string[] {
    if (!imgData) {
        return;
    }
    const bytesPerRow = Math.ceil(width / 8);
    const xbmp = new Array(height * bytesPerRow).fill(0);
    for (let y = yStart; y < yStart + height; y++) {
        for (let x = xStart; x < xStart + width; x++) {
            const imgDataIndex = (y * width + x) * 4;
            const rgbSumValue =
                imgData.data[imgDataIndex] + imgData.data[imgDataIndex + 1] + imgData.data[imgDataIndex + 2];
            const alphaValue = imgData.data[imgDataIndex + 3];
            // b&w + alpha masking
            if (alphaValue > 127 && rgbSumValue < 381) {
                const xbmpIndex = (y - yStart) * bytesPerRow + Math.floor((x - xStart) / 8);
                const bitPosition = (x - xStart) % 8;
                xbmp[xbmpIndex] |= 1 << bitPosition;
            }
        }
    }

    for (let i = 0; i < xbmp.length; i++) {
        xbmp[i] = swapBits ? bitswap(xbmp[i]) : xbmp[i];
        xbmp[i] = '0x' + xbmp[i].toString(16).padStart(2, '0');
    }

    return xbmp;
}

export function imgDataToUint32Array(imgData) {
    const length = imgData.data.length / 4; // number of pixels
    const arrayLength = Math.ceil(length / 32);
    let xbmp = new Array(arrayLength).fill(0);
    for (let y = 0; y < imgData.height; y++) {
        for (let x = 0; x < imgData.width; x++) {
            const pixelNumber = y * imgData.width + x; // Overall pixel number in the image
            const imgDataIndex = pixelNumber * 4;
            const alphaValue = imgData.data[imgDataIndex + 3];

            if (alphaValue > 127) {
                const xbmpIndex = Math.floor(pixelNumber / 32); // Index in the xbmp array
                const bitPosition = 31 - (pixelNumber % 32); // Position within the 32-bit chunk
                xbmp[xbmpIndex] |= 1 << bitPosition;
                xbmp[xbmpIndex] >>>= 0; // Convert to unsigned integer
            }
        }
    }

    return xbmp.map((x) => '0x' + x.toString(16));
}

export function toCppVariableName(str) {
    const cppKeywords = [
        'auto',
        'double',
        'int',
        'struct',
        'break',
        'else',
        'long',
        'switch',
        'case',
        'enum',
        'register',
        'typedef',
        'char',
        'extern',
        'return',
        'union',
        'const',
        'float',
        'short',
        'unsigned',
        'continue',
        'for',
        'signed',
        'void',
        'default',
        'goto',
        'sizeof',
        'volatile',
        'do',
        'if',
        'static',
        'while'
    ];

    // Replace non-alphanumeric characters with underscores
    let variableName = str.replace(/[^a-z0-9]/gi, '_');

    // Prepend an underscore if the first character is a digit
    if (/[0-9]/.test(variableName[0])) {
        variableName = '_' + variableName;
    }

    // If name is a C++ keyword, append an underscore
    if (cppKeywords.includes(variableName)) {
        variableName += '_';
    }

    return variableName;
}

export function generateUID() {
    return Math.random().toString(36).substring(2) + Date.now().toString(36);
}

export function postParentMessage(type, data) {
    if (window.top !== window.self) {
        window.top.postMessage({target: 'lopaka_app', type: type, payload: data}, '*');
    }
}

export function logEvent(event_name: string, option?: string) {
    window.gtag &&
        gtag('event', event_name, {
            app_name: 'lopaka',
            option: option
        });
}

export function throttle(func, limit = 5000) {
    let inThrottle;
    return function () {
        const args = arguments;
        const context = this;
        if (!inThrottle) {
            func.apply(context, args);
            inThrottle = true;
            setTimeout(() => (inThrottle = false), limit);
        }
    };
}

export function debounce(func, delay = 500) {
    let timeoutId;
    return function () {
        const args = arguments;
        const context = this;
        clearTimeout(timeoutId);
        timeoutId = setTimeout(() => func.apply(context, args), delay);
    };
}
