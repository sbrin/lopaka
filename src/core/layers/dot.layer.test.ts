import {expect, test, vi} from 'vitest';
import * as utils from '../../utils';
import {Point} from '../point';
import {EditMode} from './abstract.layer';
import {DotLayer} from './dot.layer';

vi.spyOn(utils, 'generateUID').mockImplementation(() => '000001');

const display = new Point(128, 64);
const scale = new Point(1, 1);

test('Create layer', () => {
    let layer = new DotLayer({
        defaultColor: '#FFFFFF',
        hasCustomFontSize: true,
        hasInvertedColors: false,
        hasRGBSupport: false,
        screenBgColor: '#000000'
    });
    layer.resize(display, scale);
    expect(layer.state).toMatchSnapshot();
    expect(layer.bounds).toMatchSnapshot();
});

test('Edit layer', () => {
    let layer = new DotLayer({
        defaultColor: '#FFFFFF',
        hasCustomFontSize: true,
        hasInvertedColors: false,
        hasRGBSupport: false,
        screenBgColor: '#000000'
    });
    layer.resize(display, scale);
    layer.startEdit(EditMode.CREATING, new Point(10, 10), layer.editPoints[0]);
    expect(layer.state).toMatchSnapshot();
    layer.edit(new Point(30, 20), new MouseEvent('mousemove'));
    layer.stopEdit();
    expect(layer.state).toMatchSnapshot();
    expect(layer.bounds).toMatchSnapshot();
});

test('Move layer', () => {
    let layer = new DotLayer({
        defaultColor: '#FFFFFF',
        hasCustomFontSize: true,
        hasInvertedColors: false,
        hasRGBSupport: false,
        screenBgColor: '#000000'
    });
    layer.resize(display, scale);
    layer.startEdit(EditMode.MOVING, new Point(10, 10), layer.editPoints[0]);
    layer.edit(new Point(30, 20), new MouseEvent('mousemove'));
    expect(layer.position.xy).toEqual([30, 20]);
    expect(layer.state).toMatchSnapshot();
    layer.edit(new Point(40, 30), new MouseEvent('mousemove'));
    expect(layer.position.xy).toEqual([40, 30]);
    layer.stopEdit();
    expect(layer.state).toMatchSnapshot();
    expect(layer.bounds).toMatchSnapshot();
});

test('Modifiers', () => {
    let layer = new DotLayer({
        defaultColor: '#FFFFFF',
        hasCustomFontSize: true,
        hasInvertedColors: true,
        hasRGBSupport: true,
        screenBgColor: '#000000'
    });
    layer.resize(display, scale);
    expect(layer.position.xy).toEqual([0, 0]);
    layer.modifiers.x.setValue(10);
    expect(layer.position.xy).toEqual([10, 0]);
    expect(layer.state).toMatchSnapshot();
    layer.modifiers.y.setValue(20);
    expect(layer.position.xy).toEqual([10, 20]);
    layer.modifiers.color.setValue('#FF0000');
    expect(layer.color).toEqual('#FF0000');
    expect(layer.state).toMatchSnapshot();
    layer.modifiers.inverted.setValue(true);
    expect(layer.state).toMatchSnapshot();
});

test('Draw layer', () => {
    let layer = new DotLayer({
        defaultColor: '#FFFFFF',
        hasCustomFontSize: true,
        hasInvertedColors: false,
        hasRGBSupport: false,
        screenBgColor: '#000000'
    });
    layer.resize(display, scale);
    layer.color = '#FF0000';
    layer.position = new Point(10, 10);
    layer.draw();
    expect((layer.getBuffer().getContext('2d') as any).getState()).toMatchSnapshot();
});
