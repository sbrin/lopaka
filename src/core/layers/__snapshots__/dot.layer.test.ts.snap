// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Create layer 1`] = `
{
  "c": "#FFFFFF",
  "f": false,
  "g": undefined,
  "i": undefined,
  "in": false,
  "n": undefined,
  "p": [
    0,
    0,
  ],
  "s": [
    1,
    1,
  ],
  "t": "dot",
  "u": "000001",
}
`;

exports[`Create layer 2`] = `
Rect {
  Symbol(pos): Point {
    Symbol(x): 0,
    Symbol(y): 0,
  },
  Symbol(size): Point {
    Symbol(x): 0,
    Symbol(y): 0,
  },
}
`;

exports[`Draw layer 1`] = `
[
  {
    "property": "fillStyle",
    "value": "#000",
  },
  {
    "property": "strokeStyle",
    "value": "#000",
  },
  {
    "args": [
      -1,
      -1,
      129,
      65,
    ],
    "command": "clearRect",
  },
  {
    "property": "fillStyle",
    "value": "#FF0000",
  },
  {
    "property": "strokeStyle",
    "value": "#FF0000",
  },
  {
    "args": [],
    "command": "beginPath",
  },
  {
    "args": [
      10,
      10,
      1,
      1,
    ],
    "command": "rect",
  },
  {
    "args": [
      10,
      10,
      1,
      1,
    ],
    "command": "rect",
  },
  {
    "args": [
      10,
      10,
      1,
      1,
    ],
    "command": "rect",
  },
  {
    "args": [
      10,
      10,
      1,
      1,
    ],
    "command": "rect",
  },
  {
    "args": [],
    "command": "fill",
  },
  {
    "args": [],
    "command": "save",
  },
  {
    "property": "fillStyle",
    "value": "rgba(0,0,0,0)",
  },
  {
    "args": [],
    "command": "beginPath",
  },
  {
    "args": [
      10,
      10,
      1,
      1,
    ],
    "command": "rect",
  },
  {
    "args": [],
    "command": "fill",
  },
  {
    "args": [],
    "command": "restore",
  },
  {
    "args": [],
    "command": "closePath",
  },
]
`;

exports[`Edit layer 1`] = `
{
  "c": "#FFFFFF",
  "f": false,
  "g": undefined,
  "i": undefined,
  "in": false,
  "n": undefined,
  "p": [
    10,
    10,
  ],
  "s": [
    1,
    1,
  ],
  "t": "dot",
  "u": "000001",
}
`;

exports[`Edit layer 2`] = `
{
  "c": "#FFFFFF",
  "f": false,
  "g": undefined,
  "i": undefined,
  "in": false,
  "n": undefined,
  "p": [
    30,
    20,
  ],
  "s": [
    1,
    1,
  ],
  "t": "dot",
  "u": "000001",
}
`;

exports[`Edit layer 3`] = `
Rect {
  Symbol(pos): Point {
    Symbol(x): 10,
    Symbol(y): 10,
  },
  Symbol(size): Point {
    Symbol(x): 1,
    Symbol(y): 1,
  },
}
`;

exports[`Modifiers 1`] = `
{
  "c": "#FFFFFF",
  "f": false,
  "g": undefined,
  "i": undefined,
  "in": false,
  "n": undefined,
  "p": [
    10,
    0,
  ],
  "s": [
    1,
    1,
  ],
  "t": "dot",
  "u": "000001",
}
`;

exports[`Modifiers 2`] = `
{
  "c": "#FF0000",
  "f": false,
  "g": undefined,
  "i": undefined,
  "in": false,
  "n": undefined,
  "p": [
    10,
    20,
  ],
  "s": [
    1,
    1,
  ],
  "t": "dot",
  "u": "000001",
}
`;

exports[`Modifiers 3`] = `
{
  "c": "#FF0000",
  "f": false,
  "g": undefined,
  "i": undefined,
  "in": true,
  "n": undefined,
  "p": [
    10,
    20,
  ],
  "s": [
    1,
    1,
  ],
  "t": "dot",
  "u": "000001",
}
`;

exports[`Move layer 1`] = `
{
  "c": "#FFFFFF",
  "f": false,
  "g": undefined,
  "i": undefined,
  "in": false,
  "n": undefined,
  "p": [
    30,
    20,
  ],
  "s": [
    1,
    1,
  ],
  "t": "dot",
  "u": "000001",
}
`;

exports[`Move layer 2`] = `
{
  "c": "#FFFFFF",
  "f": false,
  "g": undefined,
  "i": undefined,
  "in": false,
  "n": undefined,
  "p": [
    40,
    30,
  ],
  "s": [
    1,
    1,
  ],
  "t": "dot",
  "u": "000001",
}
`;

exports[`Move layer 3`] = `
Rect {
  Symbol(pos): Point {
    Symbol(x): 0,
    Symbol(y): 0,
  },
  Symbol(size): Point {
    Symbol(x): 0,
    Symbol(y): 0,
  },
}
`;
